nixos/modules/services/web-apps/youtrack.nix

```
@@ -1,131 +1,178 @@
+        See https://www.jetbrains.com/help/youtrack/server/2023.1/YouTrack-Java-Start-Parameters.html
+        for more information.
+      '';
+      example = {
+        tls-redirect-from-http = true;
+        jetbrains.youtrack.admin.restore = true;
       };
+      default = {};
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    # XXX: Drop all version feature switches at the point when we consider YT 2022.3 as outdated.
+    services.youtrack.package = lib.mkDefault (
+        if lib.versionOlder config.system.stateVersion "24.05" then pkgs.youtrack_2022_3
+        else pkgs.youtrack
+      );
+
+    services.youtrack.settings = {
+      listen-address = cfg.address;
+      listen-port = toString cfg.port;
     };
 
+    systemd.services.youtrack = let
+      service_jar = let
+        mergeAttrList = lib.foldl' lib.mergeAttrs {};
+        stdParams = mergeAttrList [
+          (lib.optionalAttrs (cfg.settings ? base-url && cfg.settings.base-url != null) {
+            "jetbrains.youtrack.baseUrl" = cfg.settings.base-url;
+          })
+          {
+          "java.aws.headless" = "true";
+          "jetbrains.youtrack.disableBrowser" = "true";
+          }
+        ];
+        extraAttr = lib.concatStringsSep " " (lib.mapAttrsToList (k: v: "-D${k}=${v}") (stdParams // cfg.settings));
+      in {
+        environment.HOME = cfg.statePath;
+        environment.YOUTRACK_JVM_OPTS = "${extraAttr}";
+        after = [ "network.target" ];
+        wantedBy = [ "multi-user.target" ];
+        path = with pkgs; [ unixtools.hostname ];
+        serviceConfig = {
+          Type = "simple";
+          User = "youtrack";
+          Group = "youtrack";
+          Restart = "on-failure";
+          ExecStart = ''${package}/bin/youtrack --J-Xmx${cfg.maxMemory} --J-XX:MaxMetaspaceSize=${cfg.maxMetaspaceSize} ${cfg.jvmOpts} ${cfg.address}:${toString cfg.port}'';
+        };
+      };
+      service_zip = {
+        after = [ "network.target" ];
+        wantedBy = [ "multi-user.target" ];
+        path = with pkgs; [ unixtools.hostname ];
+        preStart = ''
+          # This detects old (i.e. <= 2022.3) installations that were not migrated yet
+          # and migrates them to the new state directory style
+          if [[ -d ${cfg.statePath}/teamsysdata ]] && [[ ! -d ${cfg.statePath}/2022_3 ]]
+          then
+            mkdir -p ${cfg.statePath}/2022_3
+            mv ${cfg.statePath}/teamsysdata ${cfg.statePath}/2022_3
+            mv ${cfg.statePath}/.youtrack ${cfg.statePath}/2022_3
+          fi
+          mkdir -p ${cfg.statePath}/{backups,conf,data,logs,temp}
+        '' + lib.optionalString (lib.versionAtLeast package.version "2023.1") ''
+          ${package}/bin/youtrack configure -J-Ddisable.configuration.wizard.on.upgrade=true
+          ${package}/bin/youtrack configure --listen-port=${toString cfg.port} ${lib.concatStringsSep " " (lib.mapAttrsToList (name: value: "--${name}=${toString value}") cfg.settings )}
```

The port is already a part of `settings`
```suggestion
          ${package}/bin/youtrack configure ${lib.concatStringsSep " " (lib.mapAttrsToList (name: value: "--${name}=${toString value}") cfg.settings )}
```

---

Also, there is an issue with `youtrack configure` caused by the fact that the config is actually stateful, so for example, you may add `settings.tls-redirect-from-http = true;` to your configuration, switch to the configuration, remove `settings.tls-redirect-from-http = true;` from your configuration, and switch once again. However, `tls-redirect-from-http=1` will still be present in the config files within `/var/lib/youtrack/conf`

---

Sadly, AFAICS, there is no way to specify these settings as a part of `youtrack run`, so perhaps, it's worth cleaning existing configuration altogether before applying `cfg.settings`

---

That is also my level of information. I'm not sure tbh how to clean the configuration? At least AFAIK there is no file that's only written to by `youtrack confiure` and not by the web ui. 
But also the configuration is the one thing that still confuses me, so happy to accept solutions :)

---

But also I'll try that out again

---

> But also the configuration is the one thing that still confuses me, so happy to accept solutions

Agree that the configuration mechanism is now very weird, and I also didn't yet come up with a way to solve this problem:(

---

Just looked at what `youtrack configure` does exactly. When using environment parameters, it sets the value all over the `/conf` directory:
```
[root@nixpkgs:/var/lib/youtrack/conf]# rg listen-port
youtrack/service-config.properties
88:bundle.listen-port=8099

internal/services/adminService/service-config.properties
70:bundle.listen-port=8099

hub/service-config.properties
61:bundle.listen-port=8099

internal/bundle.properties
12:listen-port=8099

internal/services/startingPage/service-config.properties
66:bundle.listen-port=8099

internal/services/bundleProcess/service-config.properties
67:bundle.listen-port=8099

internal/services/configurationWizard/service-config.properties
55:jetbrains.bundle.listen-port.default=8099

internal/services/bundleBackend/service-config.properties
66:bundle.listen-port=8099
``` 

but for jvm options, it just sets them in `/conf/youtrack.jvmoptions`. And this file can probably be stored in the nix store.
```
[root@nixpkgs:/var/lib/youtrack/conf]# rg default.page
youtrack.jvmoptions
65:-Djetbrains.youtrack.default.page=/agiles/157-0/current
```

---

I now did one approach about that. Not 100% good, but much better than before imo. Feel free to review :)

---

Looks better now:)